djhtml/lines.py

```diff
@@ -2,57 +2,69 @@ class Line:
-    but now this will only happen when the source file already
-    contains a final newline.
-
-    See https://github.com/rtts/djhtml/issues/56 for the discussion
-    that led to this change.
-
     """
 
-    def __init__(self, nr=1):
-        self.nr = nr
-        self.tokens = []
-        self.level = 0
+    def __init__(self, tokens=None, level=0, offset=0, ignore=False):
+        """
+        Lines are currently never instantiated with arguments, but
+        that doesn't mean they can't.
+
+        """
+        self.tokens = tokens or []
+        self.level = level
+        self.offset = offset
+        self.ignore = ignore
 
     def append(self, token):
         """
-        Append tokens to the line.
+        Append token to line.
 
         """
-        token.line_nr = self.nr
         self.tokens.append(token)
 
     @property
     def text(self):
         """
-        The unindented text of this line without leading/trailing spaces.
+        The text of this line including the original
+        leading/trailing spaces.
+
+        """
+        return "".join([token.text for token in self.tokens])
+
+    @property
+    def indents(self):
+        """
+        Whether this line has more opening than closing tokens.
 
         """
-        return "".join([str(token) for token in self.tokens]).strip()
+        return len([token for token in self.tokens if token.indents]) > len(
+            [token for token in self.tokens if token.dedents]
+        )
 
     def indent(self, tabwidth):
         """
-        The final, indented text of this line. Make sure to set the level
-        and optionally offset before calling this method.
+        The final, indented text of this line.
 
         """
-        if self.tokens:
-            if self.tokens[0].ignore:
-                return "".join([str(token) for token in self.tokens])
-            elif self.text:
-                offset = self.tokens[0].offset * tabwidth
-                spaces = tabwidth * self.level + offset
-                return " " * spaces + self.text
+        if self.ignore:
+            return self.text
+        if text := self.text.strip():
+            return " " * (tabwidth * self.level + self.offset) + text
         return ""
 
-    def __repr__(self):
-        return repr(self.tokens)
+    def __len__(self):
+        """
+        The length of the line (so far), excluding the whitespace
+        at the beginning. Be careful calling len() because it might
+        result in trailing spaces being counted that will be removed
+        by indent().
 
-    def __bool__(self):
-        return bool(self.tokens and self.text)
+        """
+        return len(self.text.lstrip())
 
-    def __next__(self):
-        return Line(nr=self.nr + 1)
+    def __repr__(self):
+        kwargs = ""
+        for attr in ["level", "offset", "ignore"]:
+            if value := getattr(self, attr):
+                kwargs += f", {attr}={value}"
```

Comment:
I'd suggest

```
kwargs += f", {attr}={value!r}"
```

so that (if ever) you allow a str-kwarg or similar, it is at least quoted properly.